### DISCUSSION TITLE
WebSocket messages spamming

### DISCUSSION THREAD

**Discussion by mikelfried** [open-webui/open-webui#5376] [General]
# Bug Report

## Installation Method

Docker

## Environment

- **Open WebUI Version:** v0.3.21

- **Operating System:** Linux 
- **Browser (if applicable):** chrome 111

**Confirmation:**

- [v] I have read and followed all the instructions provided in the README.md.
- [v] I am on the latest version of both Open WebUI and Ollama.

## Expected Behavior:

I have a docker instance with hundreds of users, and it is very slow.
I’ve noticed that using the websocket the client receive more than 5 messages a second with a json object containing word “usage” and a the key “models” and an array of models names. It is just repeating and spamming both the server and the client.

## Actual Behavior:

Maybe send this message once every minute or every admin change

## Description

**Bug Summary:**
Web socket spamming messages

## Reproduction Details

**Steps to Reproduce:**
Just docker instance with hundreds users.

**Comment by wthrbtn** [comment]
We are experiencing the same issue in our environment:

We run Open WebUI v0.3.21 as a container in Kubernetes with a couple hundred users.

During the later part of the day, our webservice slows down to a crawl. We observe thousands of websocket requests containing usage information and the list of models.

We've noticed that some idle clients, which haven't been used for hours, have generated tens of thousands of these requests.

It appears to create a feedback loop: as more messages are sent, it triggers even more messages, exacerbating the problem over time.

**Comment by tjbck** [comment]
PR welcome here!

**Comment by wthrbtn** [comment]
I am trying to figure out the purpose of the USAGE_POOL, why are the clients informed about what models are being used?

**Comment by wthrbtn** [comment]
I notice the work with WEBSOCKET_MANAGER and WEBSOCKET_REDIS_URL :+1: 

---

I plan to test a modified backend/open_webui/apps/socket/main.py during next week in our cluster
The concept is:

main-mod.py (new additions):
1. Implement debounced broadcasting:
   Add debounced_broadcast() and schedule_broadcast()
   Limit frequency of usage updates

2. Enhance usage tracking:
   Add broadcast_usage_if_changed()
   Implement rate limiting for broadcasts

3. Modify disconnect handling:
   Clean up usage pool on disconnect
   Schedule broadcast instead of immediate emission

4. Refactor existing functions:
   Schedule broadcasts instead of immediate emissions in some cases
